gobblin-core/src/main/java/gobblin/source/extractor/extract/google/GoogleAnalyticsUnsampledExtractor.java

```diff
@@ -0,0 +1,367 @@
+                                          GoogleCommon.getJsonFactory(),
+                                          Preconditions.checkNotNull(credential, "Credential is required"))
+                                 .setApplicationName(Preconditions.checkNotNull(wuState.getProp(APPLICATION_NAME), "ApplicationName is required"))
+                                 .build();
+
+    GoogleDriveFsHelper fsHelper = closer.register(new GoogleDriveFsHelper(wuState, driveClient));
+    UnsampledReport createdReport = prepareUnsampledReport(fsHelper);
+
+    DateTime nextWatermarkDateTime = googleAnalyticsFormatter.parseDateTime(createdReport.getEndDate()).plusDays(1);
+    nextWatermark = Long.parseLong(watermarkFormatter.print(nextWatermarkDateTime));
+
+    this.actualExtractor = closer.register(new GoogleDriveExtractor<S, D>(copyOf(wuState), fsHelper));
+  }
+
+  @VisibleForTesting
+  GoogleAnalyticsUnsampledExtractor(WorkUnitState state, Extractor<S, D> actualExtractor, Analytics gaService) throws IOException {
+    this.wuState = state;
+    this.googleAnalyticsFormatter = DateTimeFormat.forPattern(DATE_FORMAT)
+        .withZone(DateTimeZone.forID(state.getProp(SOURCE_TIMEZONE, DEFAULT_SOURCE_TIMEZONE)));
+    this.watermarkFormatter = DateTimeFormat.forPattern(WATERMARK_INPUTFORMAT)
+        .withZone(DateTimeZone.forID(state.getProp(SOURCE_TIMEZONE, DEFAULT_SOURCE_TIMEZONE)));
+    this.actualExtractor = actualExtractor;
+    this.gaService = gaService;
+    this.nextWatermark = -1;
+  }
+
+  /**
+   * Copy WorkUnitState so that work unit also contains job state. FileBasedExtractor needs properties from job state (mostly source.* properties),
+   * where it has been already removed when reached here.
+   *
+   * @param src
+   * @return
+   */
+  private WorkUnitState copyOf(WorkUnitState src) {
+    WorkUnit copiedWorkUnit = WorkUnit.copyOf(src.getWorkunit());
+    copiedWorkUnit.addAllIfNotExist(src.getJobState());
+
+    WorkUnitState workUnitState = new WorkUnitState(copiedWorkUnit, src.getJobState());
+    workUnitState.addAll(src);
+    return workUnitState;
+  }
+
+  /**
+   * Create unsampled report in Google drive and add google drive file id into state so that Google drive extractor
+   * can extract record from it. Also, update the state to use CsvFileDownloader unless other downloader is defined.
+   *
+   * It also register closer to delete the file from Google Drive unless explicitly requested to not deleting it.
+   * @return documentID of unsampled report in Google drive
+   * @throws IOException
+   *
+   */
+  @VisibleForTesting
+  UnsampledReport prepareUnsampledReport(final GoogleDriveFsHelper fsHelper) throws IOException {
+    UnsampledReport createdReport = createUnsampledReports();
+
+    final String fileId = createdReport.getDriveDownloadDetails().getDocumentId();
+    LOG.info("Temporary unsampled report created in Google Drive: " + fileId);
+
+    if (wuState.getPropAsBoolean(DELETE_TEMP_UNSAMPLED_REPORT, true)) {
```

Comment:
A generic feedback: please change to retrieve all config values from state / config in constructor instead of individual methods. This is to enforce fail fast if config is incorrect and helps in